### PR TITLE
add predictor

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

add predictor

## How was this patch tested?

unit test

## Related links or issues (optional)
https://github.com/intel-analytics/BigDL/issues/524

### CODE REVIEW

**Comment by jason-dai** [comment]
Please add a Module.predict API (so that PythonBigDL.modelPredictRDD can call it).

**Comment by zhangxiaoli73** [comment]
ok, I will add

**Comment by zhangxiaoli73** [comment]
ok, I have added Module.predict API

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: should we put this in another package(e.g. predict)? It's not related with optim. @jason-dai

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.intel.analytics.bigdl._
+import com.intel.analytics.bigdl.dataset.{Sample, SampleToBatch, DataSet => _}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+abstract class Predictor[T, D](model: Module[T], dataSet: RDD[D]) extends Serializable {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: dataSet should not be a member of Predictor

Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: RDD[Sample]

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  def predict(batchSize: Int): RDD[Tensor[T]]
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: batchSize should be a part of Predictor

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  }
+}
+
+class DistriPredictor[T: ClassTag] private[optim](
+   model: Module[T],
+   dataSet: RDD[Sample[T]]
+ )(implicit ev: TensorNumeric[T]) extends Predictor[T, Sample[T]](model, dataSet) {
+
+  def predict(batchSize: Int) : RDD[Tensor[T]] = {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Input is sample but output is batch. Looks weird to me.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 5]
```diff
@@ -321,5 +323,14 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
   def getNumericType(): TensorDataType = {
     ev.getType()
   }
+
+  /**
+   * module predict
+   * @param dataset dataset for prediction
+   * @param batchSize batch size per partition
+   */
+  def predict[D](dataset: RDD[D], batchSize: Int): RDD[Tensor[T]] = {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: It should be RDD[Sample]

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+abstract class Predictor[T, D](model: Module[T], dataSet: RDD[D]) extends Serializable {
+  def predict(batchSize: Int): RDD[Tensor[T]]
+}
+
+object Predictor {
+  def apply[T: ClassTag, D](model: Module[T], dataset: RDD[D])
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: It should be RDD[Sample]

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  }
+}
+
+class DistriPredictor[T: ClassTag] private[optim](
+   model: Module[T],
+   dataSet: RDD[Sample[T]]
+ )(implicit ev: TensorNumeric[T]) extends Predictor[T, Sample[T]](model, dataSet) {
+
+  def predict(batchSize: Int) : RDD[Tensor[T]] = {
+    val modelBroadCast = dataSet.sparkContext.broadcast(
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Use Use ModelBroadcast instead; see https://github.com/intel-analytics/BigDL/blob/master/dl/src/main/scala/org/apache/spark/ml/DLClassifier.scala#L91

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+      model.evaluate(), new SampleToBatch[T](batchSize))
+    dataSet.mapPartitions { partition =>
+      val localModel = modelBroadCast.value._1.cloneModule()
+      val localTransformer = modelBroadCast.value._2.cloneTransformer()
+      val miniBatch = localTransformer(partition)
+      miniBatch.map { batch =>
+        val output = localModel.forward(batch.data).toTensor[T]
+        output
+      }
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Each record in input is a Sample; each record in output is a tensor containing multiple labels in a batch - this behavior is very confusing. You need to make sure each record in output is a label for the corresponding sample.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 1 above)
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: put in utils?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 8 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, change ongoing

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 5 above)
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: and you should change PychonBigDL.modelPredictRDD(0 to call this method.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 7 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: As I have to broadcast both model and transformer, so ModelBroadcast is not suitable.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 8 above)
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: You should broadcast them separately. Never broadcast a model and clone it directly - it can be too large.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 7 above)
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: You should broadcast them separately. never broadcast a model and clone it directly - it can be too large.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 5 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: ok, but modelPredictRDD output is Sample, not Tensor. I will disscuss with zhichao later.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 7 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, I see , thank you.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 5 above)
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: You can re-construct the sample using the input and output RDDs.

Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: ok, i will refactor the interface. the input dataset is RDD[Sample], and the output result also RDD[Sample], same to PychonBigDL.modelPredictRDD()

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 2 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, changed

Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, changed

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 4 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, output is sample now.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 6 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, changed

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 8 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, changed

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+class Predictor[T: ClassTag] private[optim](
+   model: Module[T])(implicit ev: TensorNumeric[T]) extends Serializable {
+
+  def predictClass(dataSet: RDD[Sample[T]]): RDD[Int] = {
+    val result = predict(dataSet)
+    result.mapPartitions { partition =>
+      partition.map(output => {
+        val _output = output.toTensor[T]
+        require(_output.dim() == 1, "Predictor: Only support one sample one lable ")
```
Comment (zhichao-li) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Add output.dim to the exception message

File: dl/src/test/scala/com/intel/analytics/bigdl/optim/PredictorSpec.scala [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    prob(91) should be (model.forward(data(91).feature()))
+
+    sc.stop()
+  }
+
+  "model.predictClass" should "be correct" in {
+    Engine.init(1, 1, true)
+    RNG.setSeed(100)
+    val sc = new SparkContext("local[1]", "DistriValidator")
+    val data = new Array[Sample[Float]](100)
```
Comment (zhichao-li) [inline dl/src/test/scala/com/intel/analytics/bigdl/optim/PredictorSpec.scala]: May be we can revise the unitest which the length of data is not divisible by node*core*4 to see if the result is correct or not.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 9 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, changed

File: dl/src/test/scala/com/intel/analytics/bigdl/optim/PredictorSpec.scala (hunk 10 above)
Comment (zhangxiaoli73) [inline dl/src/test/scala/com/intel/analytics/bigdl/optim/PredictorSpec.scala]: ok, changed

**Comment by zhichao-li** [comment]
any update for this?

**Comment by zhangxiaoli73** [comment]
jenkins pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala [hunk 11]
```diff
@@ -99,7 +99,8 @@ class SampleToBatch[T: ClassTag]
 (totalBatch : Int,
  featurePadding : Option[Tensor[T]] = None,
  labelPadding : Option[T] = None,
- fixedLength: Option[Int] = None)
+ fixedLength: Option[Int] = None,
+ batchCore : Option[Boolean] = None)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: I don't think we should support batchCore here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala [hunk 12]
```diff
@@ -147,7 +148,8 @@ class SampleToBatch[T: ClassTag]
     if (a > b) i else j
   }
 
-  private val batchPerCore = Utils.getBatchSize(totalBatch)
+  private val batchPerCore =
+    if (batchCore.getOrElse(false)) totalBatch else Utils.getBatchSize(totalBatch)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: I don't think we should support batchCore here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/vgg/Train.scala [hunk 13]
```diff
@@ -15,15 +15,17 @@
...
 import com.intel.analytics.bigdl.dataset.image._
-import com.intel.analytics.bigdl.nn.{ClassNLLCriterion, Module}
+import com.intel.analytics.bigdl.nn.{Module, ClassNLLCriterion}
+import com.intel.analytics.bigdl._
 import com.intel.analytics.bigdl.optim._
-import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric._
 import com.intel.analytics.bigdl.utils.{Engine, T}
 import org.apache.log4j.{Level, Logger}
 import org.apache.spark.SparkContext
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric._
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/vgg/Train.scala]: Why the changes?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+      })
+    }
+  }
+
+  def predict(dataSet: RDD[Sample[T]]): RDD[Activity] = {
+    val modelBroad = ModelBroadcast[T].broadcast(dataSet.sparkContext, model.evaluate())
+    val otherBroad = dataSet.sparkContext.broadcast(
+      new SampleToBatch[T](batchPerCore, None, None, None, Some(true)))
+    dataSet.mapPartitions { partition =>
+      val localModel = modelBroad.value().cloneModule()
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: don't call cloneModule() here

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+  private val batchPerCore = 4
+
+  def predictClass(dataSet: RDD[Sample[T]]): RDD[Int] = {
+    val result = predict(dataSet)
+    result.mapPartitions { partition =>
+      partition.map(output => {
+        val _output = output.toTensor[T]
+        require(_output.dim() == 1,
+          s"Predictor: Only support one sample one lable, but got ${_output.dim()}")
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Predictor.predictClass:

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        ev.toType[Int](_output.max(1)._2.valueAt(1))
+      })
+    }
+  }
+
+  def predict(dataSet: RDD[Sample[T]]): RDD[Activity] = {
+    val modelBroad = ModelBroadcast[T].broadcast(dataSet.sparkContext, model.evaluate())
+    val otherBroad = dataSet.sparkContext.broadcast(
+      new SampleToBatch[T](batchPerCore, None, None, None, Some(true)))
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: don't use batchPerCore

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    dataSet.mapPartitions { partition =>
+      val localModel = modelBroad.value().cloneModule()
+      val localTransformer = otherBroad.value.cloneTransformer()
+      val miniBatch = localTransformer(partition)
+      miniBatch.flatMap( batch => {
+        // todo: maybe support table
+        val output = localModel.forward(batch.data).toTensor[T]
+        output.split(1)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Why split here? It is not correct.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 15 above)
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Why do you mean by one sample one label?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala (hunk 11 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: SampleToBatch has to support both train and predict. Total batch means batch per core in predict, but means batch per node in train stage. So add a flag to distinguish.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 15 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 14 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, thanks, I have seen modelBroad.value() has done cloneModule().

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 15 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: "one sample one label" means a sample should only have one label, and then get the index of max probability  as the label.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 14 above)
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: that needs to be fixed too

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala [hunk 18]
```diff
@@ -147,7 +149,7 @@ class SampleToBatch[T: ClassTag]
     if (a > b) i else j
   }
 
-  private val batchPerCore = Utils.getBatchSize(totalBatch)
+  private val batchPerCore = Utils.getBatchSize(totalBatch, partitionNum)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: batchPerPartition

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        ev.toType[Int](_output.max(1)._2.valueAt(1))
+      })
+    }
+  }
+
+  def predict(dataSet: RDD[Sample[T]]): RDD[Activity] = {
+    val modelBroad = ModelBroadcast[T].broadcast(dataSet.sparkContext, model.evaluate())
+    val partitionNum = dataSet.partitions.length
+    val otherBroad = dataSet.sparkContext.broadcast(
+      new SampleToBatch[T](batchPerPartition * partitionNum, None, None, None, Some(partitionNum)))
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: SampleToBatch(batchSize = batchPerPartition * partitionNum, partitionNum = Some(partitionNum))

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala (hunk 18 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 19 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass
